File: solutions/haskell/bob/1/src/Bob.hs
module Bob (responseFor) where

import Data.Char

responseFor :: String -> String
responseFor x
  | isQuestion x && isYelling x = "Calm down, I know what I'm doing!"
  | isQuestion x                = "Sure."
  | isYelling x                 = "Whoa, chill out!"
  | isSilent x                  = "Fine. Be that way!"
  | otherwise                   = "Whatever."

isQuestion :: String -> Bool
isQuestion x = (elem '?' x) && (last (filter (\x -> not (isSpace x)) x) == '?')

isYelling :: String -> Bool
isYelling x = (not (null (filter isAlpha x))) && (and (map isUpper (filter isAlpha x)))

isSilent :: String -> Bool
isSilent x = and (map isSpace x)

File: solutions/cpp/bob/1/bob.cpp
#include "bob.h"

namespace bob {

    std::string hey(std::string msg) {
        bool is_question = false;
        if (msg.size() > 0) { is_question = msg.back() == '?'; }
        bool is_yelling = true;
        bool is_silence = true;
        int alpha_count = 0;
        for (size_t i = 0; i < msg.size(); ++i) {
            char c = msg.at(i);

            if (std::isalpha(c)) {
                if (is_yelling) { is_yelling = std::isupper(c); }
                ++alpha_count;
            }
            if (is_silence) { is_silence = std::isspace(c); }
        }

        if (alpha_count == 0) {
            is_yelling = false;
        }

        if (!is_silence) {
            bool done = false;
            for (size_t i = msg.size() - 1; i > 0 && !done; --i) {
                char c = msg.at(i);
                if (!std::isspace(c)) {
                    is_question = (c == '?');
                    done = true;
                }
            }
        }

        if (is_silence) {
            return "Fine. Be that way!";
        } else if (is_question && is_yelling) {
            return "Calm down, I know what I'm doing!";
        } else if (is_question) {
            return "Sure.";
        } else if (is_yelling) {
            return "Whoa, chill out!";
        } else {
            return "Whatever.";
        }
    }

}  // namespace bob
